Changing pom.xml causes modules to be compiled twice

Using https://github.com/turkeylurkey/lmp-multimodules/tree/main/testCases/typeA

1. Start multi module dev mode.
3. Make any change to `jar/pom.xml` (e.g. add whitespace somewhere)
4. Dev mode recompiles the modules twice.  This occurs every time you change pom.xml.
```
[INFO] guide-maven-multimodules-jar source compilation was successful.
[INFO] guide-maven-multimodules-jar tests compilation was successful.
[INFO] guide-maven-multimodules-war source compilation was successful.
[INFO] guide-maven-multimodules-war tests compilation was successful.
[INFO] guide-maven-multimodules-ear tests compilation was successful.
[INFO] guide-maven-multimodules-jar source compilation was successful.
[INFO] guide-maven-multimodules-jar tests compilation was successful.
[INFO] guide-maven-multimodules-war source compilation was successful.
[INFO] guide-maven-multimodules-war tests compilation was successful.
[INFO] guide-maven-multimodules-ear tests compilation was successful.
```
If there is a compile error in the source code and you change pom.xml, then the compile error is seen twice as well.
```
C:\Users\Eric\git\lmp-multimodules\testCases\typeA\jar\src\main\java\io\openliberty\guides\multimodules\lib\Converter.java:15: error: cannot find symbol
    public static int getFeet(i1nt cm) {
                              ^
  symbol:   class i1nt
  location: class Converter
1 error
[INFO] guide-maven-multimodules-jar source compilation had errors.
[INFO] guide-maven-multimodules-jar tests compilation was successful.
[INFO] guide-maven-multimodules-war source compilation was successful.
[INFO] guide-maven-multimodules-war tests compilation was successful.
[INFO] guide-maven-multimodules-ear tests compilation was successful.
C:\Users\Eric\git\lmp-multimodules\testCases\typeA\jar\src\main\java\io\openliberty\guides\multimodules\lib\Converter.java:15: error: cannot find symbol
    public static int getFeet(i1nt cm) {
                              ^
  symbol:   class i1nt
  location: class Converter
1 error
[INFO] guide-maven-multimodules-jar source compilation had errors.
[INFO] guide-maven-multimodules-jar tests compilation was successful.
[INFO] guide-maven-multimodules-war source compilation was successful.
[INFO] guide-maven-multimodules-war tests compilation was successful.
[INFO] guide-maven-multimodules-ear tests compilation was successful.
```

--- a/liberty-maven-plugin/src/it/dev-it/src/test/java/net/wasdev/wlp/test/dev/it/BaseDevTest.java
+++ b/liberty-maven-plugin/src/it/dev-it/src/test/java/net/wasdev/wlp/test/dev/it/BaseDevTest.java
@@ -309,6 +309,26 @@ public class BaseDevTest {
       return false;
    }
 
+   /**
+    * Count number of lines that contain the given string
+    */
+   protected static int countOccurrences(String str, File file) throws FileNotFoundException, IOException {
+      int occurrences = 0;
+      BufferedReader br = new BufferedReader(new FileReader(file));
+      String line = br.readLine();
+      try {
+         while (line != null) {
+            if (line.contains(str)) {
+               occurrences++;
+            }
+            line = br.readLine();
+         }
+      } finally {
+         br.close();
+      }
+      return occurrences;
+   }
+
    protected static ProcessBuilder buildProcess(String processCommand) {
       ProcessBuilder builder = new ProcessBuilder();
       builder.directory(tempProj);

--- a/liberty-maven-plugin/src/it/dev-it/src/test/java/net/wasdev/wlp/test/dev/it/MultiModuleUpdatePomsTest.java
+++ b/liberty-maven-plugin/src/it/dev-it/src/test/java/net/wasdev/wlp/test/dev/it/MultiModuleUpdatePomsTest.java
@@ -1,0 +1,109 @@
+/*******************************************************************************
+ * (c) Copyright IBM Corporation 2021.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package net.wasdev.wlp.test.dev.it;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+
+import org.apache.maven.shared.utils.io.FileUtils;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class MultiModuleUpdatePomsTest extends BaseMultiModuleTest {
+
+   @BeforeClass
+   public static void setUpBeforeClass() throws Exception {
+      setUpMultiModule("typeA", "ear", null);
+      run();
+   }
+
+   /**
+    * Test that saving a pom twice in a row only causes things to recompile once.
+    * Test updating upstream poms (jar, war) and main pom (ear).
+    */
+   @Test
+   public void updatePomsTest() throws Exception {
+
+      // Wait until the last module (ear) finishes compiling during dev mode startup
+      // TODO: this can be removed if dev mode does not recompile after first starting up
+      verifyLogMessageExists("guide-maven-multimodules-ear tests compilation was successful", 10000);
+
+      int jarSourceCount = countOccurrences("guide-maven-multimodules-jar source compilation was successful.", logFile);
+      int jarTestsCount = countOccurrences("guide-maven-multimodules-jar tests compilation was successful.", logFile);
+      int warSourceCount = countOccurrences("guide-maven-multimodules-war source compilation was successful.", logFile);
+      int warTestsCount = countOccurrences("guide-maven-multimodules-war tests compilation was successful.", logFile);
+      int earTestsCount = countOccurrences("guide-maven-multimodules-ear tests compilation was successful.", logFile);
+
+      touchFileTwice("jar/pom.xml");
+
+      // Give time for recompile to occur
+      Thread.sleep(2000);
+
+      // count exact number of messages
+      assertEquals(getLogTail(), ++jarSourceCount, countOccurrences("guide-maven-multimodules-jar source compilation was successful.", logFile));
+      assertEquals(getLogTail(), ++jarTestsCount, countOccurrences("guide-maven-multimodules-jar tests compilation was successful.", logFile));
+      assertEquals(getLogTail(), ++warSourceCount, countOccurrences("guide-maven-multimodules-war source compilation was successful.", logFile));
+      assertEquals(getLogTail(), ++warTestsCount, countOccurrences("guide-maven-multimodules-war tests compilation was successful.", logFile));
+      assertEquals(getLogTail(), ++earTestsCount, countOccurrences("guide-maven-multimodules-ear tests compilation was successful.", logFile));
+
+      touchFileTwice("war/pom.xml");
+
+      // Give time for recompile to occur
+      Thread.sleep(2000);
+
+      // count exact number of messages
+      // only war and ear should have had recompiles
+      assertEquals(getLogTail(), jarSourceCount, countOccurrences("guide-maven-multimodules-jar source compilation was successful.", logFile));
+      assertEquals(getLogTail(), jarTestsCount, countOccurrences("guide-maven-multimodules-jar tests compilation was successful.", logFile));
+      assertEquals(getLogTail(), ++warSourceCount, countOccurrences("guide-maven-multimodules-war source compilation was successful.", logFile));
+      assertEquals(getLogTail(), ++warTestsCount, countOccurrences("guide-maven-multimodules-war tests compilation was successful.", logFile));
+      assertEquals(getLogTail(), ++earTestsCount, countOccurrences("guide-maven-multimodules-ear tests compilation was successful.", logFile));
+
+      touchFileTwice("ear/pom.xml");
+
+      // Give time for recompile to occur
+      Thread.sleep(2000);
+
+      // count exact number of messages
+      // only ear should have had recompiles
+      assertEquals(getLogTail(), jarSourceCount, countOccurrences("guide-maven-multimodules-jar source compilation was successful.", logFile));
+      assertEquals(getLogTail(), jarTestsCount, countOccurrences("guide-maven-multimodules-jar tests compilation was successful.", logFile));
+      assertEquals(getLogTail(), warSourceCount, countOccurrences("guide-maven-multimodules-war source compilation was successful.", logFile));
+      assertEquals(getLogTail(), warTestsCount, countOccurrences("guide-maven-multimodules-war tests compilation was successful.", logFile));
+      assertEquals(getLogTail(), ++earTestsCount, countOccurrences("guide-maven-multimodules-ear tests compilation was successful.", logFile));
+
+
+   }
+
+private void touchFileTwice(String path) throws InterruptedException {
+      File file = new File(tempProj, path);
+      long time = System.currentTimeMillis();
+      assertTrue(file.setLastModified(time));
+      Thread.sleep(100);
+      assertTrue(file.setLastModified(time+100));
+}
+
+
+}
+

